src/java/org/apache/cassandra/service/accord/AccordTask.java

```diff
@@ -0,0 +1,880 @@
+            super(commandStore, loadCtx);
+            this.function = function;
+        }
+
+        @Override
+        public R apply(SafeCommandStore commandStore)
+        {
+            return function.apply(commandStore);
+        }
+    }
+
+    // TODO (desired): these anonymous ops are somewhat tricky to debug. We may want to at least give them names.
+    static class ForConsumer extends AccordTask<Void>
+    {
+        private final Consumer<? super SafeCommandStore> consumer;
+
+        public ForConsumer(AccordCommandStore commandStore, PreLoadContext loadCtx, Consumer<? super SafeCommandStore> consumer)
+        {
+            super(commandStore, loadCtx);
+            this.consumer = consumer;
+        }
+
+        @Override
+        public Void apply(SafeCommandStore commandStore)
+        {
+            consumer.accept(commandStore);
+            return null;
+        }
+    }
+
+    public static <T> AccordTask<T> create(CommandStore commandStore, PreLoadContext ctx, Function<? super SafeCommandStore, T> function)
+    {
+        return new ForFunction<>((AccordCommandStore) commandStore, ctx, function);
+    }
+
+    public static AccordTask<Void> create(CommandStore commandStore, PreLoadContext ctx, Consumer<? super SafeCommandStore> consumer)
+    {
+        return new ForConsumer((AccordCommandStore) commandStore, ctx, consumer);
+    }
+
+    public enum State
+    {
+        INITIALIZED(),
+        WAITING_TO_SCAN_RANGES(INITIALIZED),
+        SCANNING_RANGES(WAITING_TO_SCAN_RANGES),
+        WAITING_TO_LOAD(INITIALIZED, SCANNING_RANGES),
+        LOADING(INITIALIZED, SCANNING_RANGES, WAITING_TO_LOAD),
+        WAITING_TO_RUN(INITIALIZED, SCANNING_RANGES, WAITING_TO_LOAD, LOADING),
+        RUNNING(WAITING_TO_RUN),
+        PERSISTING(RUNNING),
+        FINISHED(RUNNING, PERSISTING),
+        CANCELLED(WAITING_TO_SCAN_RANGES, SCANNING_RANGES, WAITING_TO_LOAD, LOADING, WAITING_TO_RUN),
+        FAILED(WAITING_TO_SCAN_RANGES, SCANNING_RANGES, WAITING_TO_LOAD, LOADING, WAITING_TO_RUN, RUNNING, PERSISTING);
+
+        private final int permittedFrom;
+
+        State()
+        {
+            this.permittedFrom = 0;
+        }
+
+        State(State ... permittedFroms)
+        {
+            int permittedFrom = 0;
+            for (State state : permittedFroms)
+                permittedFrom |= 1 << state.ordinal();
+            this.permittedFrom = permittedFrom;
+        }
+
+        boolean isPermittedFrom(State prev)
+        {
+            return (permittedFrom & (1 << prev.ordinal())) != 0;
+        }
+
+        boolean isExecuted()
+        {
+            return this.compareTo(PERSISTING) >= 0;
+        }
+
+        boolean isComplete()
+        {
+            return this.compareTo(FINISHED) >= 0;
+        }
+    }
+
+    private State state = INITIALIZED;
+    private final PreLoadContext preLoadContext;
+    private final String loggingId;
+
+    @Nullable Object2ObjectHashMap<TxnId, AccordSafeCommand> commands;
```

Comment:
It can actually help the JDK to use concrete classes as it can invoke concrete methods (rather than interface methods), but I don't have a strong opinion